core/domain/prod_validation_jobs_one_off_test.py

```diff
@@ -2526,6 +2528,15 @@ def test_model_with_invalid_question_count(self):
             u'[u\'fully-validated SkillOpportunityModel\', 2]']
         run_job_and_check_output(self, expected_output, sort=True)
 
+    def test_domain_validator_failure(self):
+        self.model_instance_0.question_count = -1
+        self.model_instance_0.put()
+
+        with self.assertRaises(RuntimeError) as cm:
+            expected_output = 'Expect exception to be raised.'
```

Comment:
We should match the complete error message here. Is there any issue with that?

Reply:
Because the run_job_and_check_output() helper method runs a mapreduce job, the actual ValidateError is raised in the worker, but this manifests as a RuntimeError in the manager/master, and the validation error message is not propagated (instead it's something like 'MapReduce task to URL /mapreduce/worker/worker_callback/1560922460388testbed-request-id-hash-0 failed').

Reply:
That shouldn't be the case -- the exception raised in the validate function should be catched [here](https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off.py#L186) and rendered correctly. Further, we have such checks [here](https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off_test.py#L345) and the complete error message is matched. So, can you please look into it further and check what is the issue?

Thanks!

Reply:
Thanks for the code pointers. It looks like when the SkillOpportunityModelValidator fetches the domain objects for validation, it does so by converting the model objects into the domain objects. The init method of the domain object also called validate(), and this isn't being caught by the except block. I would expect the error to be propagated back up, but for some reason it's not. Do you have any idea why this is the case? We could remove the validate() call from the init function, but I don't think anything is implicitly wrong with calling it there.

Reply:
Ok, I get your point that the validation fails during the init part itself and so the error does not propagate. The error does not propagate since:

The code in the function [here](https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off.py#L171) which checks domain validation first fetches the domain object and then [validates the object within a try catch](https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off.py#L185). So, if there is an error in the validate function, that will be caught in the catch block and it will propagate correctly.

There is no try catch for the part where the [domain object is being fetched](https://github.com/oppia/oppia/blob/develop/core/domain/prod_validation_jobs_one_off.py#L178). So, any error there is not propagated to the final errors we store. So, my suggestion would be to move the try block as follows:

```
  try:
        model_domain_object_instance = cls._get_model_domain_object_instance(
            item)
        if model_domain_object_instance is None:
            # No domain object exists for this storage model class.
            return
        model_domain_object_instance.validate()
  except Exception as e:
         cls.errors['domain object check'].append((
                'Entity id %s: Entity fails domain validation with the '
                'error %s' % (item.id, e)))
```

Reply:
Ah, right. That's it. Thanks!

Updated the try block and the validator test.